_src/_posts/2017-01-31-learning-rtb.markdown

```diff
@@ -0,0 +1,205 @@
+that allows companies to promote their mobile applications by leveraging
+real-time bidding ([RTB][rtb]) technologies in order to buy digital advertising
+from multiple inventory sources and exchanges. During an RTB transaction, an
+auction is announced and any interested bidder has to answer with a bid price
+within a time constraint of about 100 milliseconds. The bidder that wins the
+auction pays the second-highest price and obtains the right to print a creative
+(which is just jargon for displaying an ad) on a publisher site. If the banner
+is clicked by the user further goal events might be tracked, v.g. the
+application may be installed or opened. Obtaining a goal event is called a
+conversion.
+
+One main component of our platform is the machine learning subsystem that feeds
+the bidding subsystem with predictive models of conversion rates for different
+events of interest (clicks, installs, opens, etc.). A conversion rate for an
+event is the conditional probability of getting that event given the set of
+features that characterizes the current RTB transaction and given that we
+already printed a creative targeting that event. Why is it so important for our
+business to have good estimates of these rates? Simply because we sell
+conversions but we pay for impressions. For example, suppose one client pays us
+$\$1$ per install and we estimate the install conversion rate to be around $0.1$
+(just cheating... real world figures are well under a dismaying $0.001$ `:(`).
+Then, assuming our creative is going to get printed, we should expect an average
+profit of $\$0.1 - c$, where $c$ is the expected cost of the impression. If our
+rate estimates are grossly miscalculated we will be losing money by valuating
+the expected income too low (thus missing opportunities with $c < \$0.1$) or too
+high (thus buying non-opportunities with $c > \$0.1$)[^risk]. The bottom line is
+that our bidder may be able to pick an optimal pair (bid, creative) for the
+market transaction in course from:
+(i) The estimated conditional conversion rate for the goal event.
+(ii) Our customer valuation of the goal event.
+(iii) The budget constraint for the advertising campaign.
+(iv) A narrow-minded profit maximization behavior.
+
+In order to estimate conversion rates, we implemented a SGD (Stochastic
+Gradient Descent) algorithm or, more specifically, [a FTRL-P][ftrlp] (Follow the
+Regularized Leader Proximal) algorithm with [Adagrad][adagrad] (Adaptive
+Learning Rate) and [ElasticNet][enet] (Lasso and Ridge) regularization. This is
+a state-of-the-art algorithm mostly used in online advertising because of its
+[convenient trade-off][sgd] between optimization and estimation errors in
+scenarios where computational capacity ---and not available data--- is the
+real bottleneck. Every day our system is able to learn from an online stream of
+tens of millions of events published by the bidder into a very
```

Comment:
"bidder" is the same word, but used for different contexts:
  * a bidder is a company as Jampp competing with other companies for an auction;
  * a bidder is an internal sub-system inside Jampp;
Perhaps "our bidder" is more clear than "the bidder", but a bit repetitive.